### PR TITLE
Add support for Python 3, Plone 5.2 and Plone 6

### PR DESCRIPTION
**Pull request by pbauer**

### CODE REVIEW

**Review by hvelarde — COMMENTED** [review]
sorry, guys, but it's not possible for me to review a PR with 48 commits and 106 files changed

**Comment by pbauer** [comment]
This is ready for merging.

File: src/collective/cover/tiles/edit.py [hunk 1]
```diff
@@ -57,8 +58,13 @@ def handleSave(self, action):
...
         old_data = dataManager.get()
-        for item in data:
-            old_data[item] = data[item]
+
+        form.applyChanges(self, old_data, data)
+        for group in self.groups:
+            form.applyChanges(group, old_data, data)
+
+        current_data = self.getContent()
+        current_data.update(old_data)
```
Comment (wesleybl) [inline src/collective/cover/tiles/edit.py:67]: @cleberjsantos why is this necessary?

File: src/collective/cover/tiles/file.py [hunk 2]
```diff
@@ -8,11 +8,17 @@
...
-from Products.MimetypesRegistry.common import MimeTypeException
 from zope import schema
 from zope.interface import implementer
 
 
+try:
+    from Products.MimetypesRegistry.interfaces import MimeTypeException
+except ImportError:
+    # Products.MimetypesRegistry < 2.1
+    from Products.MimetypesRegistry.common import MimeTypeException
```
Comment (wesleybl) [inline src/collective/cover/tiles/file.py:19]: This branch drop support for Plone < 5.2. So I think this can be removed.

File: src/collective/cover/tiles/list.py [hunk 3]
```diff
@@ -14,14 +14,14 @@
 from plone.tiles.interfaces import ITileDataManager
 from plone.tiles.interfaces import ITileType
 from plone.uuid.interfaces import IUUID
+from Products.CMFPlone.utils import safe_unicode
```
Comment (wesleybl) [inline src/collective/cover/tiles/list.py:17]: Use `Products.CMFPlone.utils.safe_text`

File: src/collective/cover/tiles/richtext.py [hunk 4]
```diff
@@ -12,6 +12,7 @@
 from plone.tiles.interfaces import ITileDataManager
 from plone.uuid.interfaces import IUUID
 from Products.CMFPlone.utils import safe_hasattr
+from Products.CMFPlone.utils import safe_unicode
```
Comment (wesleybl) [inline src/collective/cover/tiles/richtext.py:15]: Use `Products.CMFPlone.utils.safe_text`

File: webpack/app/galleria/theme.js [hunk 5]
```diff
@@ -12,12 +12,14 @@ var rm_button = function(data) {
   return "<i class='tile-remove-item' data-uuid='" + uuid + "'><span class='text'>remove</span></i>";
 };
 
-(function($) {
+!(function($) {
```
Comment (wesleybl) [inline webpack/app/galleria/theme.js:15]: bump version in:

https://github.com/collective/collective.cover/blob/master/webpack/package.json#L3

**Review by wesleybl — CHANGES_REQUESTED** [review]
I reviewed and left some comments.

File: src/collective/cover/tiles/list.py (hunk 3 above)
Comment (wesleybl) [reply src/collective/cover/tiles/list.py]: Done

File: src/collective/cover/tiles/richtext.py (hunk 4 above)
Comment (wesleybl) [reply src/collective/cover/tiles/richtext.py]: Done.

File: src/collective/cover/tests/test_banner_tile.py [hunk 6]
```diff
@@ -6,6 +6,7 @@
 from lxml import etree  # nosec
 from mock import Mock
 from plone.tiles.interfaces import ITileDataManager
+from Products.CMFPlone.utils import safe_unicode
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/tests/test_banner_tile.py]: Done

File: src/collective/cover/tests/test_basic_tile.py [hunk 7]
```diff
@@ -1,19 +1,20 @@
...
 from mock import Mock
 from plone import api
+from plone.app.event.base import localized_today
 from plone.app.testing import logout
 from plone.cachepurging.hooks import queuePurge
 from plone.cachepurging.interfaces import ICachePurgingSettings
 from plone.namedfile.file import NamedBlobImage
 from plone.registry.interfaces import IRegistry
 from plone.tiles.interfaces import ITileDataManager
+from Products.CMFPlone.utils import safe_unicode
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/tests/test_basic_tile.py]: Done

File: src/collective/cover/tests/test_collection_tile.py [hunk 8]
```diff
@@ -1,11 +1,12 @@
 # -*- coding: utf-8 -*-
 from collective.cover.tests.base import TestTileMixin
-from collective.cover.tests.utils import today
 from collective.cover.tiles.collection import CollectionTile
 from collective.cover.tiles.collection import ICollectionTile
 from mock import Mock
 from plone import api
+from plone.app.event.base import localized_today
 from plone.uuid.interfaces import IUUID
+from Products.CMFPlone.utils import safe_unicode
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/tests/test_collection_tile.py]: Done.

File: src/collective/cover/tests/test_file_tile.py [hunk 9]
```diff
@@ -5,6 +5,7 @@
 from collective.cover.tiles.file import IFileTile
 from plone import api
 from plone.uuid.interfaces import IUUID
+from Products.CMFPlone.utils import safe_unicode
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/tests/test_file_tile.py]: Done.

File: src/collective/cover/tiles/configuration.py [hunk 10]
```diff
@@ -2,6 +2,7 @@
 from persistent.dict import PersistentDict
 from plone.namedfile.interfaces import INamedBlobImageField
 from plone.tiles.interfaces import ITileType
+from Products.CMFPlone.utils import safe_unicode
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/tiles/configuration.py]: Done.

File: src/collective/cover/setuphandlers.py [hunk 11]
```diff
@@ -37,18 +27,5 @@ def add_default_layout():
...
-    qi.installProduct("plone.app.relationfield")
-    fti = api.portal.get_tool("portal_types")["collective.cover.content"]
-    fti.behaviors += ("plone.app.relationfield.behavior.IRelatedItems",)
-
-
 def run_after(portal_setup):
     add_default_layout()
-
-    if HAS_RELATIONFIELD:
-        install_relationfield()
```
[the remark this answers is not in the document]
Comment (pbauer) [reply src/collective/cover/setuphandlers.py:54]: relationfield is always installed in plone 5.

File: webpack/app/galleria/theme.js (hunk 5 above)
Comment (pbauer) [reply webpack/app/galleria/theme.js:15]: I bumped it from 2.1.2 to 2.1.3, is that ok?

Comment (wesleybl) [reply webpack/app/galleria/theme.js:15]: With the python package updated to 3.0.0, I think it's better to put the same version here. So we leave space if someone wants to do something on a 2.x branch.

I made it. Thanks!

File: src/collective/cover/setuphandlers.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/setuphandlers.py:54]: But you removed the `install_relationfield` function. We can remove the if but keep the function call. It adds a behavior to the cover type. It might be interesting to change the name of the function.

File: src/collective/cover/tests/test_controlpanel.py [hunk 12]
```diff
@@ -19,12 +22,14 @@ class ControlPanelTestCase(unittest.TestCase):
 
     def setUp(self):
         self.portal = self.layer["portal"]
+        self.request = self.layer["request"]
         self.controlpanel = self.portal["portal_controlpanel"]
+        self.qi = get_installer(self.portal, self.request)
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/tests/test_controlpanel.py]: Done.

File: src/collective/cover/tests/test_tiles.py [hunk 13]
```diff
@@ -23,7 +23,7 @@ def setUp(self):
     @unittest.expectedFailure  # FIXME: remove tiles on uninstall
     def test_tiles_removed_on_uninstall(self):
         qi = self.portal["portal_quickinstaller"]
-        qi.uninstallProducts(products=[PROJECTNAME])
+        qi.uninstall_product(PROJECTNAME)
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/tests/test_tiles.py:26]: Done.

File: src/collective/cover/tests/utils.py [hunk 14]
```diff
@@ -42,16 +36,15 @@ def create_standard_content_for_tests(portal):
 
         # XXX: handle setting text field for both, Archetypes and Dexterity
         set_text_field(obj, u"<p>The quick brown fox jumps over the lazy dog</p>")
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/tests/utils.py]: Done.

File: src/collective/cover/tests/utils.py [hunk 15]
```diff
@@ -2,20 +2,14 @@
 """Helper functions to create test content and work around API
 inconsistencies among Archetypes and Dexterity.
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/tests/utils.py]: Done.

File: src/collective/cover/profiles/default/metadata.xml [hunk 16]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0"?>
 <metadata>
-  <version>24</version>
+  <version>30</version>
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/profiles/default/metadata.xml:3]: Done.

File: build-i18n.sh [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+I18NPATH="src/$DOMAIN/src/collective/cover"
+
+$I18NDUDE rebuild-pot --pot ${I18NPATH}/locales/${DOMAIN}.pot --create ${DOMAIN} ${I18NPATH}
+$I18NDUDE sync --pot ${I18NPATH}/locales/${DOMAIN}.pot ${I18NPATH}/locales/*/LC_MESSAGES/${DOMAIN}.po
+
+## Don't automatically run the rebuild for the "plone" domain, as it puts too
+## much message strings into it.
+#DOMAIN=plone
+#$I18NDUDE rebuild-pot --pot $I18NPATH/locales/$DOMAIN.pot --merge $I18NPATH/locales/merge-plone.pot --create $DOMAIN $I18NPATH
+#$I18NDUDE sync --pot $I18NPATH/locales/$DOMAIN.pot $I18NPATH/locales/*/LC_MESSAGES/$DOMAIN.po
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply build-i18n.sh]: Done.

File: src/collective/cover/tests/test_controlpanel.py [hunk 18]
```diff
@@ -111,22 +117,3 @@ def test_records_removed_on_uninstall(self):
...
-        self.assertEqual(len(self.settings.layouts), 3)
-
-        qi = self.portal["portal_quickinstaller"]
-        with api.env.adopt_roles(["Manager"]):
-            qi.reinstallProducts(products=[PROJECTNAME])
-
-        self.assertIn("foo", self.settings.styles)
-        self.assertEqual(len(self.settings.styles), 5)
-        self.assertIn("Empty layout", self.settings.layouts)
-        self.assertEqual(len(self.settings.layouts), 4)
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/tests/test_controlpanel.py]: Done.

File: src/collective/cover/testing.py [hunk 19]
```diff
@@ -1,48 +1,28 @@
...
-features we want to test:
-
-plone.app.contenttypes:
-    installed under Plone 4.3, if requested; installed under Plone 5
-
-Products.PloneFormGen
-    installed under Plone 4 only (now deprecated)
-"""
-from collective.cover.config import IS_PLONE_5
+from __future__ import absolute_import
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/testing.py]: Done.

File: src/collective/cover/profiles/default/registry/bootstrap3js.xml [hunk 20]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0"?>
+<registry
+    xmlns:i18n="http://xml.zope.org/namespaces/i18n"
+    i18n:domain="collective.cover">
+
+  <records prefix="plone.bundles/bootstrap3"
+           interface='Products.CMFPlone.interfaces.IBundleRegistry'>
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/profiles/default/registry/bootstrap3js.xml]: Done.

File: src/collective/cover/setuphandlers.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (pbauer) [reply src/collective/cover/setuphandlers.py:54]: `plone.relateditems` is always enables on `collective.cover.content`. See `collective.cover.content.xml`

File: src/collective/cover/tests/test_robot.py [hunk 21]
```diff
@@ -1,17 +1,4 @@
...
-
-We have 3 kinds of expected failures:
-
-Expected Failure
-    We know they are failing and need some work to be fixed.
-Mandelbug
-    Their behaviour is chaotic and they are hard to fix.
-Issue related
-    They are failing under certain reproducible circunstances.
-"""
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/tests/test_robot.py]: Done.

File: src/collective/cover/tests/test_robot.py [hunk 22]
```diff
@@ -22,18 +9,11 @@
 
 dirname = os.path.dirname(__file__)
 files = os.listdir(dirname)
-tests = [f for f in files if f.startswith("test_") and f.endswith(".robot")]
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/tests/test_robot.py:25]: Done.

File: src/collective/cover/tests/test_views.py [hunk 23]
```diff
@@ -64,9 +64,10 @@ def test_alternate_view_render(self):
         html = etree.HTML(view())
         # alternate view should show title, description and viewlets
         self.assertEqual(len(html.xpath('.//h1[contains(text(),"My Title")]')), 1)
-        self.assertEqual(
-            len(html.xpath('.//div[contains(text(),"My Description")]')), 1
-        )
+        # FIXME: html in Plone 6 is different (<p class="lead">My Description</p>)
+        # self.assertEqual(
+        #     len(html.xpath('.//div[contains(text(),"My Description")]')), 1
+        # )
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/tests/test_views.py]: Done

File: src/collective/cover/profiles/default/registry/main.xml [hunk 24]
```diff
@@ -0,0 +1,17 @@
...
+    i18n:domain="collective.cover">
+
+  <records interface="Products.CMFPlone.interfaces.controlpanel.ITypesSchema" prefix="plone">
+    <value purge="false" key="default_page_types">
+      <element>collective.cover.content</element>
+    </value>
+
+    <value purge="false" key="displayed_types">
+      <element>collective.cover.content</element>
+    </value>
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/profiles/default/registry/main.xml]: @cleberjsantos For now I've removed it to keep it as it was before. This can be done in policy. Feel free to re-enter it if you really need it.

File: src/collective/cover/tests/utils.py [hunk 25]
```diff
@@ -42,16 +36,15 @@ def create_standard_content_for_tests(portal):
...
         api.content.create(
             container=portal,
             type="Event",
             title=u"My event",
-            startDate=DateTime(today),  # Archetypes
-            endDate=DateTime(tomorrow),
-            start=today,  # Dexterity
+            start=now,
             end=tomorrow,
-            timezone=TZNAME,
```
[the remark this answers is not in the document]
Comment (wesleybl) [reply src/collective/cover/tests/utils.py:54]: Well, the tests are passing without the timezone. So I don't think it will be needed.

File: src/collective/cover/tiles/edit.py (hunk 1 above)
Comment (wesleybl) [reply src/collective/cover/tiles/edit.py]: @cleberjsantos It's working without it. So I removed that code. Feel free to return it if needed.

**Review by wesleybl — APPROVED** [review]
LGTM.

Some adjustments still need to be made but can be done in other PRs so this one doesn't get even bigger.
